### PR TITLE
More tests for the simple_* methods

### PR DESCRIPTION
**Pull request by clokep**
After being a bit nervous about #16583, I wanted to write some tests for more of the `simple_*` methods of the `DatabasePool`.

### CODE REVIEW

File: synapse/storage/database.py [hunk 1]
```diff
@@ -2283,17 +2281,14 @@ def simple_delete_many_txn(
...
-
         clause, values = make_in_list_sql_clause(txn.database_engine, column, values)
         clauses = [clause]
 
         for key, value in keyvalues.items():
             clauses.append("%s = ?" % (key,))
             values.append(value)
 
-        if clauses:
-            sql = "%s WHERE %s" % (sql, " AND ".join(clauses))
```
Comment (clokep) [inline synapse/storage/database.py:2296]: There's always at least 1 clause, so this is just a minor clean-up.

File: tests/storage/test_base.py [hunk 2]
```diff
@@ -55,6 +55,8 @@ def runWithConnection(func, *args, **kwargs):  # type: ignore[no-untyped-def]
         engine = create_engine(sqlite_config)
```
Comment (clokep) [inline tests/storage/test_base.py:55]: I'd *love* to test these with postgres too, but not sure if there's an easy way to do that. 😢

File: tests/storage/test_base.py [hunk 3]
```diff
@@ -209,3 +209,63 @@ def test_delete_one(self) -> Generator["defer.Deferred[object]", object, None]:
         self.mock_txn.execute.assert_called_with(
             "DELETE FROM tablename WHERE keycol = ?", ["Go away"]
         )
+
+    @defer.inlineCallbacks
```
Comment (DMRobertson) [inline tests/storage/test_base.py:321]: I guess twisted/trial doesn't let you write test methods using `async def ...`? :(

File: tests/storage/test_base.py [hunk 4]
```diff
@@ -209,3 +209,63 @@ def test_delete_one(self) -> Generator["defer.Deferred[object]", object, None]:
...
+    def test_upsert(self) -> Generator["defer.Deferred[object]", object, None]:
```
Comment (DMRobertson) [inline tests/storage/test_base.py:380]: There's some stuff in synapse about "emulated upserts". Do we want to add tests for those?

(Aside: my view would be that we should try to rip out that logic. I think we would need to enforce that background updates which make the necessary unique indices have run?)

File: tests/storage/test_base.py [hunk 5]
```diff
@@ -209,3 +209,63 @@ def test_delete_one(self) -> Generator["defer.Deferred[object]", object, None]:
...
+        result = yield defer.ensureDeferred(
+            self.datastore.db_pool.simple_upsert(
+                table="tablename",
+                keyvalues={"columnname": "oldvalue"},
+                values={"othercol": "newvalue"},
+            )
+        )
+
+        self.mock_txn.execute.assert_called_with(
+            "INSERT INTO tablename (columnname, othercol) VALUES (?, ?) ON CONFLICT (columnname)  DO UPDATE SET othercol=EXCLUDED.othercol",
```
Comment (DMRobertson) [inline tests/storage/test_base.py:392]: Aside: where does the extra space between `(columnname)` and `DO UPDATE` come from? (I guess this is just newlines getting converted to a regular space?)

File: tests/storage/test_base.py [hunk 6]
```diff
@@ -269,3 +269,45 @@ def test_upsert_with_where(
...
+        yield defer.ensureDeferred(
+            self.datastore.db_pool.simple_upsert_many(
+                table="tablename",
+                key_names=["columnname"],
+                key_values=[["oldvalue"]],
+                value_names=["othercol"],
+                value_values=[["newvalue"]],
+                desc="",
+            )
+        )
```
Comment (DMRobertson) [inline tests/storage/test_base.py:450]: might be worth getting this to run an upsert which affects more than one row?

File: tests/storage/test_base.py (hunk 2 above)
Comment (DMRobertson) [reply tests/storage/test_base.py:55]: What's the objection/blocker here, sorry? Would mocking out calls to psycopg2 suffice, or is the point that you want to assert what SQL gets executed?

**Review by DMRobertson — APPROVED** [review]
Looks sensible; no objections. Not sure I understand what the difficulty is with postgres though. Shout if you want a rubber duck.

File: tests/storage/test_base.py (hunk 3 above)
Comment (clokep) [reply tests/storage/test_base.py:321]: They do, but I think not on our lowest bound. I started testing it and got into some dependency hell and stopped.

File: tests/storage/test_base.py (hunk 5 above)
Comment (clokep) [reply tests/storage/test_base.py:392]: > I guess this is just newlines getting converted to a regular space?)

It is there in-case there's a where clause.

File: tests/storage/test_base.py (hunk 2 above)
Comment (clokep) [reply tests/storage/test_base.py:55]: Just seems like a big faff to sometimes test against postgres and sometimes test against sqlite. I was having trouble figuring out how to properly structure the tests. I'll poke at it again.

File: tests/storage/test_base.py (hunk 4 above)
Comment (clokep) [reply tests/storage/test_base.py:380]: > There's some stuff in synapse about "emulated upserts". Do we want to add tests for those?

It would probably be good to do this, yeah. Agreed that it would be better to *not* have that though. 😄 See #16397 maybe?

File: tests/storage/test_base.py [hunk 7]
```diff
@@ -396,7 +397,28 @@ def test_upsert(self) -> Generator["defer.Deferred[object]", object, None]:
         self.assertTrue(result)
 
     @defer.inlineCallbacks
-    def test_upsert_with_insert(
+    def test_upsert_no_values(
```
Comment (DMRobertson) [inline tests/storage/test_base.py:457]: Are you sure this is testing the "emulation" logic? I thought that a table-level lock was required: https://github.com/matrix-org/synapse/blob/cfb6d38c47711b8dfaf0125353aec88d16708b97/synapse/storage/database.py#L1188-L1189

(maybe this is PG only since there is only ever a single sqlite writer?)

Comment (DMRobertson) [reply tests/storage/test_base.py:457]: Ahhh: https://github.com/matrix-org/synapse/blob/cfb6d38c47711b8dfaf0125353aec88d16708b97/synapse/storage/database.py#L1220-L1221

Comment (DMRobertson) [reply tests/storage/test_base.py:457]: https://github.com/matrix-org/synapse/blob/cfb6d38c47711b8dfaf0125353aec88d16708b97/synapse/storage/database.py#L1271-L1288 I think you may need to patch/mock `_unsafe_to_upsert_tables`?

Comment (clokep) [reply tests/storage/test_base.py:457]: It does -- it adds the table into the `_unsafe_to_upsert_tables` set.

Comment (clokep) [reply tests/storage/test_base.py:457]: > Are you sure this (and others) is testing the "emulation" logic? I thought that a table-level lock was required:

That gets called if `lock` is set to `True`, which is the default:

https://github.com/matrix-org/synapse/blob/80a7f75e4dc4d8612c61f02faa8b17bcc8af54d0/synapse/storage/database.py#L1315-L1317

tl;dr -- it uses a different API on the engine. For [sqlite this is a no-op](https://github.com/matrix-org/synapse/blob/4f1840a88ad3a93244fc23149c56245704eab824/synapse/storage/engines/sqlite.py#L118-L119), but for [postgres it does something](https://github.com/matrix-org/synapse/blob/4f1840a88ad3a93244fc23149c56245704eab824/synapse/storage/engines/postgres.py#L197-L198).

I think this is because [`assert_called_with`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_with) checks the *last* call only, ignoring prior calls.

Comment (clokep) [reply tests/storage/test_base.py:457]: I switched everything to `assert_called_once_with`, then fixed the Postgres-specific ones to add the lock call.

**Review by DMRobertson — APPROVED** [review]
Thanks for tolerating my pedantry :)
